// sms77-client/dist/Sms77Client.umd.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.Sms77Client = f()}})(function(){var define,module,exports;return (function(){function r(e,n,t){function o(i,f){if(!n[i]){if(!e[i]){var c="function"==typeof require&&require;if(!f&&c)return c(i,!0);if(u)return u(i,!0);var a=new Error("Cannot find module '"+i+"'");throw a.code="MODULE_NOT_FOUND",a}var p=n[i]={exports:{}};e[i][0].call(p.exports,function(r){var n=e[i][1][r];return o(n||r)},p,p.exports,r,e,n,t)}return n[i].exports}for(var u="function"==typeof require&&require,i=0;i<t.length;i++)o(t[i]);return o}return r})()({1:[function(require,module,exports){
module.exports = require('../src/Sms77Client').default;
},{"../src/Sms77Client":3}],2:[function(require,module,exports){
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
var ERROR_CODES_1 = require("./constants/ERROR_CODES");
var BaseClient = /** @class */ (function () {
    function BaseClient(apiKey, sendWith) {
        this.apiKey = apiKey;
        this.sendWith = sendWith;
    }
    BaseClient.prototype.post = function (endpoint, data) {
        return __awaiter(this, void 0, void 0, function () {
            var params, boolToNumber, queryString, cfg, res, text, code;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        params = __assign(__assign({}, data), { p: this.apiKey, sendWith: this.sendWith });
                        boolToNumber = function (v) { return 'boolean' === typeof v ? v ? 1 : 0 : v; };
                        queryString = Object.entries(params).map(function (_a) {
                            var k = _a[0], v = _a[1];
                            return encodeURIComponent(k) + "=" + encodeURIComponent(boolToNumber(v));
                        }).join('&');
                        cfg = {
                            headers: {
                                'Content-Type': 'multipart/form-data'
                            },
                            method: 'POST'
                        };
                        return [4 /*yield*/, fetch(BaseClient.BASE_URL + "/" + endpoint + "?" + queryString, cfg)];
                    case 1:
                        res = _a.sent();
                        return [4 /*yield*/, res.text()];
                    case 2:
                        text = _a.sent();
                        code = Number.parseInt(text);
                        if (ERROR_CODES_1.ERROR_CODES.has(code)) {
                            throw new Error(code + ": " + ERROR_CODES_1.ERROR_CODES.get(code));
                        }
                        try {
                            return [2 /*return*/, JSON.parse(text)];
                        }
                        catch (e) {
                        }
                        return [2 /*return*/, text];
                }
            });
        });
    };
    BaseClient.BASE_URL = 'https://gateway.sms77.io/api';
    return BaseClient;
}());
exports["default"] = BaseClient;

},{"./constants/ERROR_CODES":4}],3:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
__exportStar(require("./types"), exports);
var SUCCESS_CODES_1 = require("./constants/SUCCESS_CODES");
var BaseClient_1 = __importDefault(require("./BaseClient"));
var splitByLine = function (str) { return str.split('\n'); };
var Sms77Client = /** @class */ (function (_super) {
    __extends(Sms77Client, _super);
    function Sms77Client(apiKey, sendWith) {
        if (sendWith === void 0) { sendWith = 'js'; }
        var _this = _super.call(this, apiKey, sendWith) || this;
        _this.apiKey = apiKey;
        _this.sendWith = sendWith;
        return _this;
    }
    Sms77Client.prototype.balance = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _b = (_a = Number).parseFloat;
                        return [4 /*yield*/, this.post('balance')];
                    case 1: return [2 /*return*/, _b.apply(_a, [_c.sent()])];
                }
            });
        });
    };
    Sms77Client.prototype.contacts = function (p) {
        return __awaiter(this, void 0, void 0, function () {
            var res;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.post('contacts', p)];
                    case 1:
                        res = _a.sent();
                        if ('string' === typeof res) { //return CSV if ContactsParams.json was set to 0
                            return [2 /*return*/, res];
                        }
                        else if (Array.isArray(res)) {
                            return [2 /*return*/, res];
                        }
                        if (SUCCESS_CODES_1.SUCCESS_CODES.has(Number.parseInt(res["return"]))) {
                            if (res.id) {
                                return [2 /*return*/, Number.parseInt(res.id)];
                            }
                            else if (p.id) {
                                return [2 /*return*/, p.id];
                            }
                        }
                        throw new Error(res);
                }
            });
        });
    };
    Sms77Client.prototype.lookup = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var res, isValidResponse;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.post('lookup', params)];
                    case 1:
                        res = _a.sent();
                        isValidResponse = function (str) { return 100 === Number.parseInt(str); };
                        if ('string' === typeof res) {
                            if (!isValidResponse(res)) {
                                throw new Error(res);
                            }
                        }
                        else if (res.code) {
                            if (!isValidResponse(res.code)) {
                                throw new Error(res);
                            }
                        }
                        return [2 /*return*/, res];
                }
            });
        });
    };
    Sms77Client.prototype.pricing = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.post('pricing', params)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Sms77Client.prototype.sms = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.post('sms', params)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Sms77Client.prototype.status = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var res, _a, report, timestamp;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.post('status', params)];
                    case 1:
                        res = _b.sent();
                        _a = splitByLine(res), report = _a[0], timestamp = _a[1];
                        return [2 /*return*/, { report: report, timestamp: timestamp }];
                }
            });
        });
    };
    Sms77Client.prototype.validateForVoice = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.post('validate_for_voice', params)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Sms77Client.prototype.voice = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var res, _a, code, id, cost;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.post('voice', params)];
                    case 1:
                        res = _b.sent();
                        _a = splitByLine(res), code = _a[0], id = _a[1], cost = _a[2];
                        return [2 /*return*/, {
                                code: Number.parseInt(code),
                                id: Number.parseInt(id),
                                cost: Number.parseFloat(cost)
                            }];
                }
            });
        });
    };
    return Sms77Client;
}(BaseClient_1["default"]));
exports["default"] = Sms77Client;

},{"./BaseClient":2,"./constants/SUCCESS_CODES":5,"./types":6}],4:[function(require,module,exports){
"use strict";
exports.__esModule = true;
exports.ERROR_CODES = void 0;
exports.ERROR_CODES = new Map([
    [201, 'Country code invalid.'],
    [202, 'Recipient number invalid.'],
    [300, 'Authentication missing.'],
    [301, 'Parameter to missing.'],
    [304, 'Parameter type missing.'],
    [305, 'Parameter text missing.'],
    [306, 'Sender invalid.'],
    [307, 'Parameter url missing.'],
    [400, 'Invalid type.'],
    [401, 'Parameter text too long.'],
    [402, 'Reload lock.'],
    [500, 'Not enough credits.'],
    [600, 'Carrier failed.'],
    [700, 'Unknown error.'],
    [801, 'Logo file missing.'],
    [802, 'Logo file not existent.'],
    [803, 'Ring tone missing.'],
    [900, 'Wrong API key!'],
    [901, 'Wrong message ID.'],
    [902, 'API deactivated.'],
    [903, 'IP not allowed.'],
]);

},{}],5:[function(require,module,exports){
"use strict";
exports.__esModule = true;
exports.SUCCESS_CODES = void 0;
exports.SUCCESS_CODES = new Map([
    [151, 'Contacts updated/created/deleted with CSV response.'],
    [152, 'Contacts updated/created/deleted with JSON response.'],
]);

},{}],6:[function(require,module,exports){
"use strict";
exports.__esModule = true;

},{}]},{},[1])(1)
});
